mri/generators/base.py

```diff
@@ -0,0 +1,94 @@
+"""
+Base Class for KspaceGenerators.
+
+Kspace Generators emulates the acquisition of MRI data, by providing sequential access to data.
+"""
+
+import numpy as np
+import progressbar
+
+
+class KspaceGeneratorBase:
+    """
+    Basic K-space Generator emulate the acquisition of an MRI.
+
+    K-space generator are regular Python generators, with extra function to access data property.
+
+    At each iteration the relevant kspace data and mask is returned.
+
+    Parameters
+    ----------
+    full_kspace: np.ndarray
+        The fully sampled kspace, which will be returned incrementally,
+        use for the Fourier transform.
+    mask: np.ndarray
+        A binary mask, giving the sampled location for the kspace
+    """
+
+    def __init__(self, full_kspace: np.ndarray, mask: np.ndarray, max_iter: int = 1):
+        """Create a Basic k-space generator.
+
+        Parameters
+        -----------
+        full_kspace: ndarray
+            The full kspace data
+        mask: ndarray
+            The mask for undersampling the k-space data.
+        max_iter: int
+            Maximum number of iterations to be yields.
+        """
+        self._full_kspace = full_kspace
+        self.kspace = full_kspace.copy()
+        self.mask = mask
+        self._len = max_iter
+        self.iter = 0
+
+    @property
+    def shape(self):
+        """Return shape of full_kspace."""
+        return self._full_kspace.shape
+
+    @property
+    def dtype(self):
+        """Return dtype of full_kspace."""
+        return self._full_kspace.dtype
+
+    def __len__(self):
+        """The Number of kspace acquisitions available."""
+        return self._len
+
+    def __iter__(self):
+        return self
+
+    def __getitem__(self, idx):
+        if idx >= self._len:
+            raise IndexError
+        return self._full_kspace, self.mask
+
+    def __next__(self):
+        if self.iter < self._len:
+            self.iter += 1
+            return self._full_kspace, self.mask
+        raise StopIteration
+
+    def reset(self):
+        """Reset the Generator to its initial state."""
+        self.iter = 0
+
+    def opt_iterate(self, opt, reset=True, estimate_call_period=None):
+        x_new_list = []
+        if reset:
+            self.reset()
+        for (kspace, col) in progressbar.progressbar(self):
+            opt.idx += 1
+            opt._grad.obs_data = kspace
+            opt._grad.fourier_op.mask = col
+            opt._update()
+            if opt.metrics and opt.metric_call_period is not None:
+                if opt.idx % opt.metric_call_period == 0 or opt.idx == (self._len - 1):
+                    opt._compute_metrics()
+            if estimate_call_period is not None:
+                if opt.idx % estimate_call_period == 0 or opt.idx == (self._len - 1):
+                    x_new_list.append(opt.get_notify_observers_kwargs()["x_new"])
```

Comment:
Isnt this similar to what is done in modopt? Isnt it redundant?

Reply:
Forward-Backward and primal-dual algorithms in modopt indeed have  an `iterate` methods. 
Here the idea is that it is the generator, which drive the reconstruction, as it is this class that yields the data required for it. This code could also be  in `run_online_algorithm` (see PR #134), but I find it more relevant here. One could for instance easily modify this method to provide an another way of iterating on the data (for instance doing a pre-processing step before calling `opt._update`).

Reply:
How about something like:
```python
opt._grad.obs_data = kspace
opt._grad.fourier_op.mask = col
opt.iterate(num_iterations=1) # 1 or any number of iterations, to me it must be an argument that can be set. 
                                               # In reality we want as many iterations as need till we get next k-space data
```

That will reduce a lot of the redundant codes.. 

```